src/calliope/config.py

```diff
@@ -0,0 +1,278 @@
+
+    ensure_feasibility: bool = Field(default=False)
+    """
+    Whether to include decision variables in the model which will meet unmet demand or consume unused supply in the model so that the optimisation solves successfully.
+    This should only be used as a debugging option (as any unmet demand/unused supply is a sign of improper model formulation).
+    """
+
+    objective: str = Field(default="min_cost_optimisation")
+    """Name of internal objective function to use, from those defined in the pre-defined math and any applied additional math."""
+
+    pre_validate_math_strings: bool = Field(default=True)
+    """
+    If true, the Calliope math definition will be scanned for parsing errors _before_ undertaking the much more expensive operation of building the optimisation problem.
+    You can switch this off (e.g., if you know there are no parsing errors) to reduce overall build time.
+    """
+
+    operate: BuildOperate = BuildOperate()
+
+
+class SolveSpores(ConfigBaseModel):
+    """SPORES configuration options used when solving a Calliope optimisation problem (`calliope.Model.solve`)."""
+
+    model_config = {"title": "Model solve SPORES mode configuration"}
+    number: int = Field(default=3)
+    """SPORES mode number of iterations after the initial base run."""
+
+    score_cost_class: str = Field(default="score")
+    """SPORES mode cost class to vary between iterations after the initial base run."""
+
+    slack_cost_group: str = Field(default="monetary")
+    """SPORES mode cost class to keep below the given `slack` (usually "monetary")."""
+
+    save_per_spore: bool = Field(default=False)
+    """
+    Whether or not to save the result of each SPORES mode run between iterations.
+    If False, will consolidate all iterations into one dataset after completion of N iterations (defined by `number`) and save that one dataset.
+    """
+
+    save_per_spore_path: str | None = Field(default=None)
+    """If saving per spore, the path to save to."""
+
+    skip_cost_op: bool = Field(default=False)
+    """If the model already contains `plan` mode results, use those as the initial base run results and start with SPORES iterations immediately."""
+
+    @model_validator(mode="after")
+    def require_save_per_spore_path(self) -> Self:
+        """Ensure that path is given if saving per spore."""
+        if self.save_per_spore:
+            if self.save_per_spore_path is None:
+                raise ValueError(
+                    "Must define `save_per_spore_path` if you want to save each SPORES result separately."
+                )
+            elif not Path(self.save_per_spore_path).is_dir():
+                raise ValueError("`save_per_spore_path` must be a directory.")
+        return self
+
+
+class Solve(ConfigBaseModel):
+    """Base configuration options used when solving a Calliope optimisation problem (`calliope.Model.solve`)."""
+
+    model_config = {"title": "Model Solve Configuration"}
```

Comment:
@brynpickering one of the reasons I removed these `model_config=...` was that they override the settings given in `ConfigBaseModel`. If we do want these titles, we have to copy-paste the following in all cases. The titles did not seem worth the extra lines, and that's why I took them out.

```python
    model_config = {
        "extra": "forbid",
        "frozen": True,
        "revalidate_instances": "always",
        "use_attribute_docstrings": True,
    }
```

Reply:
Oddly enough, they don't override the `ConfigBaseModel` config. You can check it out for yourself (or maybe we should add a test to ensure that remains the case)!

Reply:
Huh. Good to know!
I looked into the older version of the code, and it seems like the cause was that one of the subclasses omitted "frozen". I assumed that meant you had to specify it every time.